lhotse/features/whisper_fbank.py

```diff
@@ -0,0 +1,171 @@
+from dataclasses import dataclass
+from typing import Any, Dict, Optional, Union
+
+import numpy as np
+import torch
+
+from lhotse.features.base import FeatureExtractor, register_extractor
+from lhotse.utils import (
+    EPSILON,
+    Seconds,
+    asdict_nonull,
+    compute_num_frames_from_samples,
+    is_module_available,
+)
+
+
+def log_mel_spectrogram(
+    audio: Union[np.ndarray, torch.Tensor],
+    n_mels: int = 80,
+    n_fft: int = 400,
+    hop_length: int = 160,
+    sampling_rate: int = 16000,
+    device: Optional[Union[str, torch.device]] = None,
+):
+    """
+    From https://github.com/openai/whisper/blob/main/whisper/audio.py
+
+    Compute the log-Mel spectrogram of
+
+    Parameters
+    ----------
+    audio: Union[str, np.ndarray, torch.Tensor], shape = (*)
+        The path to audio or either a NumPy array or Tensor containing the audio waveform in 16 kHz
+
+    n_mels: int
+        The number of Mel-frequency filters, only 80 is supported
+
+    padding: int
+        Number of zero samples to pad to the right
+
+    device: Optional[Union[str, torch.device]]
+        If given, the audio tensor is moved to this device before STFT
+
+    Returns
+    -------
+    torch.Tensor, shape = (n_frames, 80)
+        A Tensor that contains the Mel spectrogram
+    """
+    if is_module_available("librosa"):
+        import librosa
+    else:
+        raise ImportError(
+            "Librosa is not installed. Please install librosa before using LibrosaFbank extractor."
+        )
+    if not torch.is_tensor(audio):
+        audio = torch.from_numpy(audio)
+
+    if device is not None:
+        audio = audio.to(device)
+    audio = audio.squeeze(0)
+    window = torch.hann_window(n_fft).to(audio.device)
+    stft = torch.stft(audio, n_fft, hop_length, window=window, return_complex=True)
+    magnitudes = stft[..., :-1].abs() ** 2
+
+    filters = librosa.filters.mel(sr=sampling_rate, n_fft=n_fft, n_mels=n_mels)
+    filters = torch.from_numpy(filters).to(device)
+    mel_spec = filters @ magnitudes
+
+    log_spec = torch.clamp(mel_spec, min=1e-10).log10()
+    log_spec = torch.maximum(log_spec, log_spec.max() - 8.0)
+    log_spec = (log_spec + 4.0) / 4.0
+
+    padding = compute_num_frames_from_samples(
+        num_samples=len(audio),
+        frame_shift=hop_length / sampling_rate,
+        sampling_rate=sampling_rate,
+    )
+    if padding > log_spec.shape[1]:
+        log_spec = torch.nn.functional.pad(
+            log_spec, (0, padding - log_spec.shape[1]), mode="constant"
+        )
+    # change shape from 80, n_frames to n_frames,80
+    log_spec = log_spec.transpose(0, 1)
+
+    return log_spec
+
+
+@dataclass
+class WhisperFbankConfig:
+    sampling_rate: int = 16000
+    num_filters: int = 80
+    hop_length: int = 160
+    n_fft: int = 400
```

Comment:
Whisper only supports these values and no other values can be used. Shall we hard-code them and don't make them configurable?